pkg/app/mutate_cluster_app.go

```diff
@@ -0,0 +1,159 @@
+	if app.Spec.Version != "" {
+		currentClusterAppVersion, err := semver.NewVersion(app.Spec.Version)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		if !currentClusterAppVersion.GreaterThan(lastClusterAppVersionWithoutRelease) {
+			// This is a cluster that is using an older version of the cluster-$provider app, a version that does
+			// not use new releases, so it does not have release version. Therefore, we don't overwrite cluster app
+			// version here.
+			//
+			// True for cluster-aws versions older or equal to v0.76.1.
+			return nil, nil
+		}
+
+		// Additional temporary check during testing.
+		//
+		// Check if this is a dev build on top of the latest version.
+		isNewDevBuild := currentClusterAppVersion.Major() == lastClusterAppVersionWithoutRelease.Major() &&
+			currentClusterAppVersion.Minor() == lastClusterAppVersionWithoutRelease.Minor() &&
+			currentClusterAppVersion.Patch() == lastClusterAppVersionWithoutRelease.Patch() &&
+			currentClusterAppVersion.Prerelease() != ""
+		// The last "standalone" cluster-aws version will be v0.76.1, but current dev builds on top of it are per semver
+		// actually older, and here we want to treat custom dev build with Releases as newer.
+		clusterAppReleasesDevBuild := semver.New(0, 76, 1, "b76af2c26f4224ffb0d718e940e232fac05c89a0", "")
+		isClusterAppReleasesDevBuild := currentClusterAppVersion.Equal(clusterAppReleasesDevBuild)
+
+		if isNewDevBuild && !isClusterAppReleasesDevBuild {
+			// This is a development cluster that is using a dev version of the cluster-$provider app that does not use
+			// new releases, so it does not have release version. Therefore, we don't overwrite cluster app
+			// version here.
+			return nil, nil
+		}
+	}
+
+	userConfigMapName := app.Spec.UserConfig.ConfigMap.Name
+	if userConfigMapName == "" {
+		return nil, microerror.Maskf(clusterAppUserConfigNotSet, "Cluster App '%s/%s does not have the user config", app.Namespace, app.Name)
+	}
+	userConfigMapNameSpace := app.Spec.UserConfig.ConfigMap.Namespace
+	if userConfigMapNameSpace == "" {
+		userConfigMapNameSpace = app.Namespace
+	}
+
+	// Now let's get the release resource from which we can read the cluster-$provider App version
+
+	// First we need user config, to get release version from it.
+	var userValuesConfigMap *v1.ConfigMap
+	getUserValues := func() error {
+		var err error
+		userValuesConfigMap, err = m.k8sClient.K8sClient().CoreV1().ConfigMaps(userConfigMapNameSpace).Get(ctx, userConfigMapName, metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		return nil
+	}
+	b := backoff.NewConstant(15*time.Second, 5*time.Second)
+	n := backoff.NewNotifier(m.logger, ctx)
+	err := backoff.RetryNotify(getUserValues, b, n)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var userValues struct {
+		Global struct {
+			Release struct {
+				Version string `yaml:"version"`
+			} `yaml:"release"`
+		} `yaml:"global"`
+	}
```

Comment:
do we have the struct in a library so could we use it from there?

Reply:
Not yet, but we will need it soon as this is escalating 🙈